docs/RFCS/time_series_culling.md

```diff
@@ -0,0 +1,349 @@
+quickly determine if each range contains time series keys (by inspecting 
+start/end keys); for ranges that do contain time series keys, specific keys
+can then be inspected at the engine level. This means that key inspections do
+not require network calls, and the number of keys that can be inspected at once
+is limited to the size of a range.
+
+Once the queue discovers a range that contains time series keys, the scanning
+process does not need to inspect every key on the range. The algorithm is as
+follows:
+
+1. Find the first time series key in the range (scan for [ts prefix]). 
+2. Deconstruct the key to retrieve its name.
+3. Run the rollup/delete operation on all keys in the range: 
+    `[ts prefix][series name][0] - [tsprefix][series name][now - threshold]`
+4. Find the next key on the range which contains data for a different time
+series by searching for key `PrefixEnd([ts prefix][series name])`.
+5. If a key was found in step 5, return to step 2 with that series name.
+
+This algorithm will avoid scanning keys that do not need to be rolled up; this
+is desirable, as once the culling algorithm is in place and has run once, the
+majority of time series keys will *not* need to be culled.
+
+The queue will be configured to run only on the range leader for a given range
+in order to avoid duplicate work; however, this is *not* necessary for
+correctness, as demonstrated in the [Rollup Algorithm](#rollup-algorithm)
+section below.
+
+The queue will initially be set to process replicas at the same rate as the
+replica GC queue (as of this RFC, one range per 50 milliseconds).
+
+##### Package Dependency
+
+There is one particular complication to this method: *go package dependency*.
+Knowledge on how to identify and cull time series keys is contained in the `ts`
+package, but all logic for replica queues (and all current queues) lives in
+`storage`, meaning that one of three things must happen:
+
++ `storage` can depend on `ts`. This seems to be trivially possible now, but may
+be unintuitive to those trying to understand our code-base. For reference, the
+`storage` package used to depend on the `sql` package in order to record event
+logs, but this eventually became an impediment to new development and had to be
+modified.
++ The queue logic could be implemented in `ts`, and `storage` could implement
+an interface that allows it to use the `ts` code without a dependency.
++ Parts of the `ts` package could be split off into another package that can
+intuitively live below `storage`. However, this is likely to be a considerable
+portion of `ts` in order to properly implement rollups.
+
+Tenatively, we will be attempting to use the first method and have `storage`
+depend on `ts`; if it is indeed trivially possible, this will be the fastest
+method of completing this project.
+
+#### Culling low resolution data
+
+Although the volume is much lower, low-resolution data will still build
```

Comment:
Those are essentially folded up into the exact same mechanism where the culling of an unculled series results in a single culled sample at the new resolution, and the culling of a culled one results in zero samples generated, right?